osbuild/buildroot.py

```diff
@@ -14,115 +19,169 @@
 ]
 
 
+# pylint: disable=too-many-instance-attributes
 class BuildRoot(contextlib.AbstractContextManager):
+    """Build Root
+
+    This class implements a context-manager that maintains a root file-system
+    for contained environments. When entering the context, the required
+    file-system setup is performed, and it is automatically torn down when
+    exiting.
+
+    The `run()` method allows running applications in this environment. Some
+    state is persistent across runs, including data in `/var`. It is deleted
+    only when exiting the context manager.
+    """
+
+    _exitstack = None
+    _rootdir = None
+    _rundir = None
+    _vardir = None
+    _libdir = None
+    _runner = None
+
+    # Paths to the directories used in the build-root.
+    api = None
+    dev = None
+    var = None
```

Comment:
Same as above. I kinda feel instance variables should be declared at the instance level (`__init__`). Or is there anything gained from having `Buildroot.api` (`None`)? With this pattern one essentially has two `api`, the one at the class level, which gets then shadowed by the instance one when it is being set. I am not strongly against it, just feels weird to me, and I am not sure what is gained by it; maybe I am missing something?

Reply:
Ah, I once ran into issues where `pylint` complained about use of undefined variables (e.g., if you use variables from `__del__()`, then `__init__()` did not necessarily run). This is when I started that. But I guess this is all obsolete. I moved it to `__init__()`, which is definitely more python-esque.